ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -704,6 +707,17 @@ chainSelectionForBlock cdb@CDB{..} blockCache hdr punish = electric $ do
         curTip      = castPoint $ AF.headPoint curChain
         curHead     = AF.headAnchor curChain
 
+        -- Either frag extends loe or loe extends frag
+        followsLoEFrag :: LoE (AnchoredFragment (Header blk))
+                      -> AnchoredFragment (Header blk)
+                      -> Bool
+        followsLoEFrag LoEDisabled _ = True
+        followsLoEFrag (LoEEnabled loe) frag =
+          case AF.intersect frag loe of
+            Just (_, _, AF.Empty{}, _) -> True
+            Just (_, _, _, AF.Empty{}) -> True
+            _                          -> False
```

Comment:
This has complexity $O(\log(n_2)n_1)$ where $n_1,n_2$ are the lengths of `frag` and `loe`, resp. I think implementing this as
```suggestion
             AF.withinFragmentBounds (AF.headPoint loe) frag
          || AF.withinFragmentBounds (AF.headPoint frag) loe
```
should also work and have complexity $O(\log(\max(n_1,n_2))$.

Also: maybe extract this (or rather the core logic) to `Ouroboros.Consensus.Util.AnchoredFragment` as sth like `onSameChain` with preconditions and even tests?